File: Arboles/dfs/dfs-list-ady.cpp
#include <vector>
#include <map>
#include <string>
#include <cstdio>
#include <iostream>

using namespace std;

int MAX = 10000;
int n, t = 0;
vector<vector<int>> adj(MAX);
vector<int> visitado(MAX), pred(MAX), f(MAX), d(MAX);

// 0 = no visitado, 1 = visitado pero no completado, 2 = completado
void dfsAux(int w){
  int u, i;
  //d[v] = ++t;
  visitado[w] = 1;
  printf("%d\n", w);

  for(i = 0; i < adj[w].size(); ++i){
    u = adj[w][i];
    if(visitado[u] == 0){
      //pred[u] = w;
      dfsAux(u);
    }
  }

  //visitado[v] = 2;
  //f[v] = ++t;
}

void dfs(){
  int i;
  
  for(i = 0; i < n; ++i){
    visitado[i] = 0;
    //pred[i] = -1;
  }
  
  for(i = 0; i < n; ++i){
    if(visitado[i] == 0)
      dfsAux(i);
  }
}

int main(){
  int m, u, v, i, j;
  map<string, int> id;

  cin >> n >> m;

  for(i = 0; i < n; ++i){
    cin >> u;
    adj[i].resize(u);
    for(j = 0; j < u; ++j)
      cin >> adj[i][j];
  }

  printf("Grafo\n");
  for(i = 0; i < n; ++i){
    printf("Nodo %d:", i);
    for(j = 0; j < adj[i].size(); ++j)
      printf(" %d", adj[i][j]);
    printf("\n");
  }

  printf("Recorrido en profundidad grafo:\n");
  dfs();
  printf("Predecesores:\n");
  for(i = 0; i < n; ++i)
    printf("El predecesor de %d es %d\n", i, pred[i]);

  return 0;
}

File: Arboles/dfs/dfs-list-ady.py
from sys import stdin

MAX = 10000
adj = [[] for i in range(MAX)]
visitado = [0 for i in range(MAX)]
p = [-1 for i in range(MAX)]
d, f = [-1 for _ in range(MAX)], [-1 for _ in range(MAX)]
n, t = int(), 0

# 0 = no visitado, 1 = visitado pero no completado, 2 = completado
def dfsAux(v):
    d[v] = t
    t += 1
    visitado[v] = 1
    print(v)

    for u in adj[v]:
        if visitado[u] == 0:
            p[u] = v
            dfsAux(u)
    f[v] = t
    t += 1
    #visitado[v] = 2

def dfs():
    for i in range(n):
        visitado[i] = 0
        p[i] = -1

    for i in range(n):
        if visitado[i] == 0:
            dfsAux(i)

def main():
    global n
    n, m = list(map(int, stdin.readline().split()))

    for i in range(n):
        vals = list(map(int, stdin.readline().split()))
        u = vals[0]
        for j in range(u):
            adj[i].append(vals[j + 1])

    print("Grafo")
    for i in range(n):
        print("Nodo %d:" % i)
        for j in range(len(adj[i])):
            print("%d" % adj[i][j], end = ' ')
        print("")

    print("Recorrido en profundidad grafo:")
    dfs()
    print("Predecesores:")
    for i in range(n):
        print("El predecesor de %d es %d" % (i, pred[i]))

main()

    
    
